pkg/types/types.go

```diff
@@ -202,9 +202,11 @@ type DeviceProvider interface {
 	GetDiscoveredDevices() []*ghw.PCIDevice
 
 	// GetDevices runs through the Discovered Devices and returns a list of fully populated HostDevices according to the given ResourceConfig
-	GetDevices(*ResourceConfig) []HostDevice
+	GetDevices(*ResourceConfig, int) []HostDevice
```

Comment:
you could replace the int (index) parameter with interface{} passing in the selectorObj.
then you will not need to deal with index handling throughout the stack. this is also relevant for GetFIlteredDevices below.

An alternative way would be to have a single "DeviceConfig" struct which will contain all needed information and it would be passed around from device provider to device implementation (pciNetDevice, AuxNetDevice etc...)

Thoughts ?

personally i prefer the latter as we would then decouple user facing API from internal one.

Reply:
Hi Adrian,

I wish to understand your second idea in more detail. 

Right now, the `resourceManager` reads the config & forwards it to the device plugins. Are you suggesting that the `resourceManager` converts the config from a `resourceConfig` style object to a `deviceConfig`, where `deviceConfig` then gets piped through the stack, like the index currently does?

For this change, a single `resourceConfig` would likely generate multiple `deviceConfigs` (one for each selector) in order to avoid the same index issue, where parts of the stack need to be told which selector/data to read. Is this what you had in mind? Or is there a simpler approach I am not seeing?

I like the idea of decoupling user-facing changes from internal changes. If we decide to go this route, we will have to decide if that work should live in a separate PR, and if so, which PR should be completed first. I think a separate PR for separating the internal and user-facing APIs without any functionality changes makes sense.